### PR TITLE
gRPC Gateway Removal

### LINKED ISSUE #12682: Remove dependency on forked gRPC gateway
# 🚀 Feature Request

### Description

For several years, Prysm has used a fork of gRPC gateway with a replace statement in go.mod. The original context can be found in https://github.com/prysmaticlabs/prysm/pull/8697. The gist of how that context relates to this issue is that the fork of grpc-gateway has some crude limited support for certain cast types in Prysm as well as some proto definitions that probably could be included here within the prysm repo. Additionally, there seems to be some changes with regards to event streaming.

See the diff between the upstream grpc-gateway and Prysm's fork. https://github.com/grpc-ecosystem/grpc-gateway/compare/main...prysmaticlabs:grpc-gateway:master

### Describe the solution you'd like

We need to evaluate if we need this fork dependency now that the JSON API is extremely different than using grpc-gateway as we did many years ago. It appears as if we can remove this dependency on the fork of grpc-gateway and move any relevant protos into Prysm's proto directory as needed. 

- [ ] Remove dependency on Prysm's forked grpc-gateway
- [ ] Move still-relevant protos into prysm
- [ ] Ensure no breaking changes with regards to the events API streaming service (grpc and json via grpc gateway) or plan deprecation of the events stream via grpc-gateway. 

### Describe alternatives you've considered

None

### PR DESCRIPTION
**Pull request by james-prysm**
- [x] I have read [CONTRIBUTING.md](CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

**What type of PR is this?**

Cleanup

**What does this PR do? Why is it needed?**

- rename GRPC flags to HTTP flags
  - the `disable-grpc-gateway` flag will be deprecated and removed in a future release.
  The following flags are being renamed:
  - `grpc-gateway-host` is renamed to `http-host`. The old name can still be used as an alias.
  - `grpc-gateway-port` is renamed to `http-port`.
  - `grpc-gateway-corsdomain` is renamed to `http-cors-domain`.
- removes all gateway mentions in code
- fully removes support of HTTP based gRPC endpoints ( internal endpoints) `breaking change`
- removes unneeded configuration around HTTP server
- removes `most` of gateway dependencies 
- removes validator gRPC service entirely

GRPC gateway dependency still in code due to the following dependencies
go.opencensus.io@v0.18.0 github.com/grpc-ecosystem/grpc-gateway@v1.5.0
go.etcd.io/etcd@v0.0.0-20191023171146-3cf2f69b5738 github.com/grpc-ecosystem/grpc-gateway@v1.9.5 <- used by Prometheus dependency

## **BREAKING CHANGES**

The HTTP implementation of gRPC endpoints has been removed as part of this PR. This was never openly recommended to be used and was primarily used internally with our systems.
 **ONLY EFFECTS THE HTTP VERSION** 
The gRPC implementation of these endpoints still exist and continue to work for the time being. 
They are listed below and are prefixed with `eth/v1alpha1`
- `/eth/v1alpha1/node/syncing`
- `/node/peers`
- `/eth/v1alpha1/node/genesis` 
- `/eth/v1alpha1/node/version`
- `/eth/v1alpha1/node/services`
- `/eth/v1alpha1/node/p2p` 
- `/eth/v1alpha1/node/peer`
- `/eth/v1alpha1/node/eth1/connections`
- `/eth/v1alpha1/validator/duties`   
- `/eth/v1alpha1/validator/duties/stream` 
- `/eth/v1alpha1/validator/domain`
- `/eth/v1alpha1/validator/chainstart/stream`
- `/eth/v1alpha1/validator/activation/stream`
- `/eth/v1alpha1/validator/index`
- `/eth/v1alpha1/validator/status`
- `/eth/v1alpha1/validator/statuses`
- `/eth/v1alpha2/validator/block` GET,POST
- `/eth/v1alpha1/validator/prepare_beacon_proposer`  GET,POST
- `/eth/v1alpha1/validator/fee_recipient_by_pub_key`
- `/eth/v1alpha1/validator/attestation` GET,POST
- `/eth/v1alpha1/validator/aggregate`  GET,POST
- `/eth/v1alpha1/validator/exit`
- `/eth/v1alpha1/validator/subnet/subscribe`
- `/eth/v1alpha1/validator/doppelganger`
- `/eth/v1alpha1/validator/sync_message_block_root`   
- `/eth/v1alpha1/validator/sync_message`
- `/eth/v1alpha1/sync_subcommittee_index`
- `/eth/v1alpha1/validator/contribution_and_proof`
- `/eth/v1alpha1/validator/signed_contribution_and_proof`
- `/eth/v1alpha1/validator/blocks/stream`
- `/eth/v1alpha1/validator/registration`
- `/eth/v1alpha1/slasher/attestations/slashable`
- `/eth/v1alpha1/slasher/blocks/slashable`
- `/eth/v1alpha1/slasher/attestations/highest`
- `/eth/v1alpha1/health/logs/stream`
- `/eth/v1alpha1/debug/state`
- `/eth/v1alpha1/debug/block`
- `/eth/v1alpha1/debug/logging`
- `/eth/v1alpha1/debug/peers`
- `/eth/v1alpha1/debug/peer`
- `/eth/v1alpha1/debug/inclusion`
- `/eth/v1alpha1/beacon/attestations`
- `/eth/v1alpha1/beacon/attestations/indexed`
- `/eth/v1alpha1/beacon/attestations/stream`
- `/eth/v1alpha1/beacon/attestations/indexed/stream`
- `/eth/v1alpha1/beacon/attestations/pool`
- `/eth/v1alpha2/beacon/blocks`
- `/eth/v1alpha1/beacon/blocks/stream`
- `/eth/v1alpha1/beacon/chainhead/stream`
- `/eth/v1alpha1/beacon/chainhead`
- `/eth/v1alpha1/beacon/committees`
- `/eth/v1alpha1/validators/balances`
- `/eth/v1alpha1/validators` GET, POST
- `/eth/v1alpha1/validators/activesetchanges`
- `/eth/v1alpha1/validators/queue`
- `/eth/v1alpha1/validators/performance`
- `/eth/v1alpha1/validators/assignments`
- `/eth/v1alpha1/validators/participation`
- `/eth/v1alpha1/beacon/config`
- `/eth/v1alpha1/beacon/validators/info/stream`
- `/eth/v1alpha1/beacon/slashings/attester/submit`
- `/eth/v1alpha1/beacon/slashings/proposer/submit`
- `/eth/v1alpha1/beacon/individual_votes`                                 

**PLEASE USE THE BEACON API REST ENDPOINTS INSTEAD**
https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/
if the HTTP REST version is not available the API is for internal use only and can only be accessed via gRPC

## Tested
- Web sanity checks with use of API endpoints
- E2E via TestEndToEnd_MinimalConfig_ValidatorRESTApi

**Which issues(s) does this PR fix?**

Fixes #12682

**Other notes for review**

### CODE REVIEW

File: deps.bzl [hunk 1]
```diff
@@ -1833,12 +1765,6 @@ def prysm_deps():
         sum = "h1:ujPKutqRlJtcfWk6toYVYagwra7HQHbXOaS171b4Tg8=",
         version = "v0.0.0-20150330215556-f50fe3d243e1",
     )
-    go_repository(
```
Comment (james-prysm) [inline deps.bzl:1836]: this got removed... should this be? need help reviewing all this

Comment (prestonvanloon) [reply deps.bzl:1836]: this is mostly generated config. Pruning unused deps is OK

File: api/server/http-rest/server.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+type Server struct {
+	cfg          *config
+	server       *http.Server
+	cancel       context.CancelFunc
+	ctx          context.Context
+	startFailure error
+}
+
+// New returns a new instance of the Server.
+func New(ctx context.Context, opts ...Option) (*Server, error) {
```
Comment (prestonvanloon) [inline api/server/http-rest/server.go:42]: What is this package about? seems unrelated to removing gRPC gateway

Comment (james-prysm) [reply api/server/http-rest/server.go:42]: this is the rename/rewrite of api/gateway/gateway.go the rest server still needs to run

Comment (prestonvanloon) [reply api/server/http-rest/server.go:42]: These are all showing up as new additions, that was my confusion. Is it too difficult to go back and move these files via `git mv` so that they are linked as a rename/move rather than an entirely new package? That would make it easier to review the diff

Comment (james-prysm) [reply api/server/http-rest/server.go:42]: need to double check, might be difficult to do that, I didn't realize git mv did that 😓 
the contents of the gateway package got moved to this http-rest package, and the internals removed. I think that's the main difference, can attempt to do it later today

Comment (james-prysm) [reply api/server/http-rest/server.go:42]: i tried to git mv but didn't work out correctly

I created a temp branch from this one
reverted the branch to develop ( maybe I should have force pushed?)
moved the files, committed
merged the temp branch back to this one 
pushed. 

it didn't seem to work.

File: api/server/http-rest/options.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+func WithHTTPAddr(addr string) Option {
+	return func(g *Server) error {
+		g.cfg.httpAddr = addr
+		return nil
+	}
+}
+
+// WithRouter allows adding a custom mux router to the gateway.
```
Comment (james-prysm) [inline api/server/http-rest/options.go:25]: need to update this, removing gateway wording, same with comments below

File: cmd/beacon-chain/flags/base.go [hunk 4]
```diff
@@ -100,30 +100,33 @@ var (
 		Usage: "Comma-separated list of API module names. Possible values: `" + PrysmAPIModule + `,` + EthAPIModule + "`.",
 		Value: PrysmAPIModule + `,` + EthAPIModule,
 	}
-	// DisableGRPCGateway for JSON-HTTP requests to the beacon node.
-	DisableGRPCGateway = &cli.BoolFlag{
+	// Deprecated: DeprecatedDisableGRPCGateway for JSON-HTTP requests to the beacon node.
+	DeprecatedDisableGRPCGateway = &cli.BoolFlag{
```
Comment (james-prysm) [inline cmd/beacon-chain/flags/base.go:104]: It's not used anywhere, there isn't a disable gateway anyways, wonder if I should deprecate or just remove

File: cmd/beacon-chain/flags/base.go [hunk 5]
```diff
@@ -100,30 +100,33 @@ var (
...
-		Name: "grpc-gateway-corsdomain",
-		Usage: "Comma separated list of domains from which to accept cross origin requests " +
-			"(browser enforced). This flag has no effect if not used with --grpc-gateway-port.",
-		Value: "http://localhost:4200,http://localhost:7500,http://127.0.0.1:4200,http://127.0.0.1:7500,http://0.0.0.0:4200,http://0.0.0.0:7500,http://localhost:3000,http://0.0.0.0:3000,http://127.0.0.1:3000",
+	// HTTPServerHost specifies a HTTP server host for the validator client.
+	HTTPServerHost = &cli.StringFlag{
+		Name:    "http-host",
+		Usage:   "Host on which the HTTP server runs on.",
+		Value:   "127.0.0.1",
+		Aliases: []string{"grpc-gateway-host"},
```
Comment (james-prysm) [inline cmd/beacon-chain/flags/base.go:113]: using aliases for this should be fine right? instead of deprecating

File: beacon-chain/rpc/eth/helpers/sync.go [hunk 6]
```diff
@@ -8,58 +8,15 @@ import (
...
 	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
 	"github.com/prysmaticlabs/prysm/v5/encoding/bytesutil"
 	"github.com/prysmaticlabs/prysm/v5/time/slots"
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
 )
 
-// ValidateSyncGRPC checks whether the node is currently syncing and returns an error if it is.
-// It also appends syncing info to gRPC headers.
-func ValidateSyncGRPC(
```
Comment (james-prysm) [inline beacon-chain/rpc/eth/helpers/sync.go:27]: I don't think this is used anywhere

File: api/server/http-rest/server.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			return nil, err
+		}
+	}
+
+	if g.cfg.router == nil {
+		return nil, errors.New("router option not configured")
+	}
+
+	corsMux := middleware.CorsHandler(g.cfg.allowedOrigins).Middleware(g.cfg.router)
+	// TODO: actually use the timeout config provided
```
Comment (saolyn) [inline api/server/http-rest/server.go:58]: Is this something that is to be added in this PR?

File: cmd/beacon-chain/flags/base.go (hunk 5 above)
Comment (saolyn) [reply cmd/beacon-chain/flags/base.go:113]: I would think so, since we want the transition from gRPC to HTTP to be pretty easy

File: cmd/beacon-chain/flags/base.go (hunk 4 above)
Comment (saolyn) [reply cmd/beacon-chain/flags/base.go:104]: Since we have the flag could we use it to disableGRPC instead so that we can use it to toggle between HTTP and GRPC when we are finally testing it's replacement?

File: api/server/http-rest/server.go (hunk 2 above)
Comment (saolyn) [reply api/server/http-rest/server.go:42]: I opened up a second window with the PR and just compared them side by side

File: api/server/http-rest/server.go (hunk 7 above)
Comment (james-prysm) [reply api/server/http-rest/server.go:58]: Yeah this is a reminder we need to add a feature for this it wasn't actually used not sure when it may have been used but definitely broken for some time

File: cmd/beacon-chain/flags/base.go (hunk 5 above)
Comment (saolyn) [reply cmd/beacon-chain/flags/base.go:113]: I think this should be fine since we want the transition between GRPC and HTTP to be painless

File: cmd/beacon-chain/flags/base.go [hunk 8]
```diff
@@ -100,30 +100,33 @@ var (
 		Usage: "Comma-separated list of API module names. Possible values: `" + PrysmAPIModule + `,` + EthAPIModule + "`.",
 		Value: PrysmAPIModule + `,` + EthAPIModule,
 	}
-	// DisableGRPCGateway for JSON-HTTP requests to the beacon node.
-	DisableGRPCGateway = &cli.BoolFlag{
+	// Deprecated: DeprecatedDisableGRPCGateway for JSON-HTTP requests to the beacon node.
+	DeprecatedDisableGRPCGateway = &cli.BoolFlag{
 		Name:  "disable-grpc-gateway",
 		Usage: "Disable the gRPC gateway for JSON-HTTP requests",
```
Comment (rkapka) [inline cmd/beacon-chain/flags/base.go:106]:
```suggestion
		Usage: "Flag deprecated and unused",
```

File: testing/endtoend/helpers/helpers.go [hunk 9]
```diff
@@ -132,7 +132,7 @@ func WaitForTextInFile(src *os.File, match string) error {
 
 	select {
 	case <-ctx.Done():
-		return fmt.Errorf("could not find requested text \"%s\" in %s before deadline:\n", match, f.Name())
```
Comment (rkapka) [inline testing/endtoend/helpers/helpers.go:135]: why change this?

File: validator/package/validator.yaml [hunk 10]
```diff
@@ -15,7 +15,7 @@ wallet-dir: /var/lib/prysm/validator
 # beacon-rpc-provider: Beacon node RPC provider endpoint. Default: localhost:4000
 # rpc-host: Specify the RPC host exposed by the validator. Default: localhost
 # rpc-port: Specify the RPC port exposed by the validator. Default: 7000
-# grpc-gateway-host: Specify the gRPC gateway port exposed by the validator. Default: localhost
-# grpc-gateway-port: Specify the gRPC gateway port exposed by the validator. Default: 7500
+# http-host: Specify the http host exposed by the validator. Default: localhost
```
Comment (rkapka) [inline validator/package/validator.yaml:18]:
```suggestion
# http-host: Specify the HTTP host exposed by the validator. Default: localhost
```

File: validator/package/validator.yaml [hunk 11]
```diff
@@ -15,7 +15,7 @@ wallet-dir: /var/lib/prysm/validator
...
+# http-port: Specify the http port exposed by the validator. Default: 7500
```
Comment (rkapka) [inline validator/package/validator.yaml:19]:
```suggestion
# http-port: Specify the HTTP port exposed by the validator. Default: 7500
```

File: api/server/http-rest/options.go (hunk 3 above)
Comment (rkapka) [reply api/server/http-rest/options.go:25]: Currently I am of the opinion that self-explanatory function names don't need to be documented. A simple `// WithRouter --` suffices to get no warnings.

File: api/server/http-rest/server.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+// Config parameters for setting up the http-rest service.
+type config struct {
+	httpAddr       string
+	allowedOrigins []string
+	muxHandler     restHandler
+	router         *mux.Router
+	timeout        time.Duration
+}
+
+// Server serves HTTP JSON traffic.
```
Comment (rkapka) [inline api/server/http-rest/server.go:32]:
```suggestion
// Server serves HTTP JSON traffic.
```
That's not entirely true because we also serve SSZ. As mentioned earlier, I don't see a good reason to document this unless there is something non-obvious that you want to explain.

File: api/server/http-rest/server.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	corsMux := middleware.CorsHandler(g.cfg.allowedOrigins).Middleware(g.cfg.router)
+	// TODO: actually use the timeout config provided
+	g.server = &http.Server{
+		Addr:              g.cfg.httpAddr,
+		Handler:           corsMux,
+		ReadHeaderTimeout: time.Second,
+	}
+	if g.cfg.muxHandler != nil { // rest APIS and Web UI registration
```
Comment (rkapka) [inline api/server/http-rest/server.go:64]: This comment is weirdly placed

Comment (james-prysm) [reply api/server/http-rest/server.go:64]: you know what this is not even needed lol good reminder

File: api/gateway/gateway.go [hunk 14]
```diff
@@ -1,212 +0,0 @@
...
-	cfg          *config
-	conn         *grpc.ClientConn
-	server       *http.Server
-	cancel       context.CancelFunc
-	ctx          context.Context
-	startFailure error
-}
-
-// New returns a new instance of the Gateway.
-func New(ctx context.Context, opts ...Option) (*Gateway, error) {
```
Comment (james-prysm) [inline api/gateway/gateway.go:65]: This got moved to api/server/httprest package

File: api/server/httprest/server.go [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/runtime"
+)
+
+var _ runtime.Service = (*Server)(nil)
+
+// httpHandler is a functional interface that implements the http handler functionality.
+type httpHandler func(
```
Comment (james-prysm) [inline api/server/httprest/server.go:16]: Need to see if I can just remove this outright or import directly into beacon node package for use

File: api/server/httprest/server.go [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+		if err := opt(g); err != nil {
+			return nil, err
+		}
+	}
+
+	if g.cfg.router == nil {
+		return nil, errors.New("router option not configured")
+	}
+
+	// TODO: actually use the timeout config provided
```
Comment (rkapka) [inline api/server/httprest/server.go:55]: What about this TODO?

Comment (james-prysm) [reply api/server/httprest/server.go:55]: It's an existing issue, for some reason we just never had a way to provide a configuration for this via flags... probably should have something.

**Review by rkapka — DISMISSED** [review]

File: go.mod [hunk 17]
```diff
@@ -32,9 +32,8 @@ require (
 	github.com/google/uuid v1.4.0
 	github.com/gorilla/mux v1.8.0
 	github.com/gostaticanalysis/comment v1.4.2
-	github.com/grpc-ecosystem/go-grpc-middleware v1.2.2
```
Comment (james-prysm) [inline go.mod:35]: this got decreased in version, that shouldn't be right, right?

Comment (james-prysm) [reply go.mod:35]: i should probably revert this...

File: deps.bzl [hunk 18]
```diff
@@ -2214,8 +2134,8 @@ def prysm_deps():
     go_repository(
         name = "com_github_magiconair_properties",
         importpath = "github.com/magiconair/properties",
-        sum = "h1:ZC2Vc7/ZFkGmsVC9KvOjumD+G5lXy2RtTKyzRKO2BQ4=",
-        version = "v1.8.1",
+        sum = "h1:SesWF0c8l/IKQX0NlsED38qoBhUpneg5HIHNdy5LyEE=",
+        version = "v1.7.4-0.20170902060319-8d7837e64d3c",
```
Comment (james-prysm) [inline deps.bzl:2138]: many of these dependency versions got reduced, I wonder why

Comment (james-prysm) [reply deps.bzl:2138]: it's used with com_github_golang_gddo which hasn't changed so possibly a higher version was required for gateway

File: go.mod [hunk 19]
```diff
@@ -94,7 +93,7 @@ require (
 	golang.org/x/sync v0.7.0
 	golang.org/x/tools v0.21.0
 	google.golang.org/genproto v0.0.0-20230410155749-daa745c078e1
-	google.golang.org/grpc v1.56.3
+	google.golang.org/grpc v1.64.0
```
Comment (james-prysm) [inline go.mod:96]: should be ok to upgrade these right?

Comment (james-prysm) [reply go.mod]: got it removed

File: api/server/httprest/log.go [hunk 20]
```diff
@@ -0,0 +1,5 @@
+package httprest
+
+import "github.com/sirupsen/logrus"
+
+var log = logrus.WithField("prefix", "REST")
```
Comment (prestonvanloon) [inline api/server/httprest/log.go:5]: please no yelling

```suggestion
var log = logrus.WithField("prefix", "httprest")
```

File: testing/endtoend/helpers/helpers.go (hunk 9 above)
Comment (james-prysm) [reply testing/endtoend/helpers/helpers.go:135]: new line is not needed at the end. this is not a log

File: deps.bzl [hunk 21]
```diff
@@ -1488,20 +1434,6 @@ def prysm_deps():
...
         version = "v1.9.5",
     )
-    go_repository(
-        name = "com_github_grpc_ecosystem_grpc_gateway_v2",
-        importpath = "github.com/grpc-ecosystem/grpc-gateway/v2",
-        patch_args = ["-p1"],
-        patches = [
-            "//third_party:com_github_grpc_ecosystem_grpc_gateway_v2.patch",
-            "//third_party:com_github_grpc_ecosystem_grpc_gateway_v2_fix_emptypb.patch",
-            "//third_party:com_github_grpc_ecosystem_grpc_gateway_v2_prysm_v5.patch",
```
Comment (prestonvanloon) [inline deps.bzl:1498]: Can you delete these patches from the repo please? They are in the third_party dir

File: cmd/validator/main.go [hunk 22]
```diff
@@ -50,7 +50,7 @@ func startNode(ctx *cli.Context) error {
 
 var appFlags = []cli.Flag{
 	flags.BeaconRPCProviderFlag,
-	flags.BeaconRPCGatewayProviderFlag,
+	flags.DeprecatedBeaconRPCGatewayProviderFlag,
```
Comment (prestonvanloon) [inline cmd/validator/main.go:53]: Moved to deprecated section please

File: api/server/httprest/server.go (hunk 16 above)
Comment (prestonvanloon) [reply api/server/httprest/server.go:55]: Whats the deal with this? What timeout config? Why can't you use it?

File: api/server/httprest/server.go [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+		ReadHeaderTimeout: time.Second,
+	}
+
+	return g, nil
+}
+
+// Start the http rest service.
+func (g *Server) Start() {
+	_, cancel := context.WithCancel(g.ctx)
+	g.cancel = cancel
```
Comment (prestonvanloon) [inline api/server/httprest/server.go:68]: I don't think this is being used as intended. This cancel function will only cancel the new context which was thrown away.

Try this:

```suggestion
	g.ctx, g.cancel := context.WithCancel(g.ctx)
```

File: api/server/httprest/server_test.go [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+	address := fmt.Sprintf("%s:%d", host, port)
+
+	opts := []Option{
+		WithHTTPAddr(address),
+		WithMuxHandler(func(
+			_ http.HandlerFunc,
+			_ http.ResponseWriter,
+			_ *http.Request,
+		) {
+		}),
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply api/server/httprest/server_test.go:38]: actually should probably just remove this ...

File: api/server/httprest/server.go (hunk 16 above)
Comment (james-prysm) [reply api/server/httprest/server.go:55]: I should either remove it now or add it I guess, it's some existing config that I migrated to this implementation, but we never had a way to configure the value ( it was accepting this field previously but it was never used)

Comment (james-prysm) [reply api/server/httprest/server.go:55]: nvm i'm just going to try to set it now

File: cmd/beacon-chain/flags/base.go (hunk 4 above)
Comment (james-prysm) [reply cmd/beacon-chain/flags/base.go:122]: will move this out and into feature

File: cmd/beacon-chain/main.go [hunk 25]
```diff
@@ -49,10 +49,10 @@ var appFlags = []cli.Flag{
 	flags.CertFlag,
 	flags.KeyFlag,
 	flags.HTTPModules,
-	flags.DisableGRPCGateway,
-	flags.GRPCGatewayHost,
-	flags.GRPCGatewayPort,
-	flags.GPRCGatewayCorsDomain,
+	flags.DeprecatedDisableGRPCGateway,
```
Comment (james-prysm) [inline cmd/beacon-chain/main.go:52]: will move this

File: cmd/validator/flags/flags.go [hunk 26]
```diff
@@ -109,25 +104,27 @@ var (
...
+		Name:    "http-host",
+		Usage:   "Host on which the HTTP server runs on.",
+		Value:   DefaultHTTPServerHost,
+		Aliases: []string{"grpc-gateway-host"},
+	}
+	// HTTPServerPort enables a HTTP server port to be exposed for the validator client.
+	HTTPServerPort = &cli.IntFlag{
+		Name:    "http-port",
+		Usage:   "Port on which the HTTP server runs on.",
+		Value:   7500,
```
Comment (saolyn) [inline cmd/validator/flags/flags.go:118]: Should we have a variable for this: `DefaultHTTPServerPort` like we do for the host

File: cmd/validator/flags/flags.go [hunk 27]
```diff
@@ -109,25 +104,27 @@ var (
...
+		Aliases: []string{"grpc-gateway-port"},
+	}
+	// HTTPServerCorsDomain adds accepted cross origin request addresses.
+	HTTPServerCorsDomain = &cli.StringFlag{
+		Name:    "corsdomain",
+		Usage:   `Comma separated list of domains from which to accept cross origin requests (browser enforced).`,
+		Value:   "http://localhost:7500,http://127.0.0.1:7500,http://0.0.0.0:7500,http://localhost:4242,http://127.0.0.1:4242,http://localhost:4200,http://0.0.0.0:4242,http://127.0.0.1:4200,http://0.0.0.0:4200,http://localhost:3000,http://0.0.0.0:3000,http://127.0.0.1:3000",
```
Comment (saolyn) [inline cmd/validator/flags/flags.go:125]: Same for this? It should improve readability ever so slightly

File: cmd/validator/flags/flags.go (hunk 26 above)
Comment (james-prysm) [reply cmd/validator/flags/flags.go:118]: sure that seems like a good idea

File: cmd/validator/flags/flags.go (hunk 27 above)
Comment (james-prysm) [reply cmd/validator/flags/flags.go:125]: like name each of these urls?

File: CHANGELOG.md [hunk 28]
```diff
@@ -23,13 +23,16 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
...
+- `grpc-gateway-corsdomain` is renamed to http-cors-domain.
 
 ### Deprecated
-
+- `--disable-grpc-gateway` flag is deprecated due to grpc gateway removal.
 - `--enable-experimental-state` flag is deprecated. This feature is now on by default. Opt-out with `--disable-experimental-state`.
 
 ### Removed
-
+- removed gRPC Gateway and gRPC versions of REST endpoints ( unused endpoints )
```
Comment (rkapka) [inline CHANGELOG.md:35]:
```suggestion
- removed gRPC Gateway
```
I would remove the second part as it is not really true. What we are really removing is the option to use REST for gRPC endpoints. And I am not sure what you mean by "unused" - someone might be still using them.

File: api/server/httprest/options.go [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+type Option func(g *Server) error
+
+func WithMuxHandler(m httpHandler) Option {
+	return func(g *Server) error {
+		g.cfg.handler = m
+		return nil
+	}
+}
+
+func WithHTTPAddr(addr string) Option {
```
Comment (rkapka) [inline api/server/httprest/options.go:18]: Missing doc

File: api/server/httprest/options.go [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+type Option func(g *Server) error
+
+func WithMuxHandler(m httpHandler) Option {
```
Comment (rkapka) [inline api/server/httprest/options.go:11]: Missing doc

File: api/server/httprest/options.go [hunk 31]
```diff
@@ -0,0 +1,39 @@
+package httprest
+
+import (
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+type Option func(g *Server) error
```
Comment (rkapka) [inline api/server/httprest/options.go:9]: Missing doc

File: config/features/deprecated_flags.go [hunk 32]
```diff
@@ -57,6 +59,20 @@ var (
 		Usage:  deprecatedUsage,
 		Hidden: true,
 	}
+
+	deprecatedBeaconRPCGatewayProviderFlag = &cli.StringFlag{
+		Name:   "beacon-rpc-gateway-provider",
+		Usage:  deprecatedUsage,
+		Value:  "127.0.0.1:3500",
```
Comment (rkapka) [inline config/features/deprecated_flags.go:66]: What's the point of having `Value` in a deprecated flag?

File: validator/rpc/server.go [hunk 33]
```diff
@@ -6,34 +6,27 @@ import (
...
 
-// Config options for the gRPC server.
+// Config options for the HTTP server.
 type Config struct {
-	Host                   string
-	Port                   string
-	GRPCGatewayHost        string
-	GRPCGatewayPort        int
+	Host                   string // http host, not grpc
+	Port                   int    // http port, not grpc
```
Comment (rkapka) [inline validator/rpc/server.go:29]: If you name the fields `HTTPHost` and `HTTPPort`, there will be no need for the comments.

File: validator/rpc/server.go [hunk 34]
```diff
@@ -51,20 +44,17 @@ type Config struct {
...
 type Server struct {
 	ctx                       context.Context
 	cancel                    context.CancelFunc
-	host                      string
-	port                      string
-	grpcGatewayHost           string
-	grpcGatewayPort           int
-	listener                  net.Listener
+	host                      string // http host, not grpc
+	port                      int    // http port, not grpc
```
Comment (rkapka) [inline validator/rpc/server.go:52]: If you name the fields `httpHost` and `httpPort`, there will be no need for the comments.

File: config/features/deprecated_flags.go (hunk 32 above)
Comment (james-prysm) [reply config/features/deprecated_flags.go:66]: i guess no real point ...

File: cmd/validator/flags/flags.go (hunk 27 above)
Comment (saolyn) [reply cmd/validator/flags/flags.go:125]: Think it would be okay to just have a cont for the entire string, at least that way it's not in the middle of the code but can be modified from the top. Making them each individual values would be more of a pain:

`const DefaultHTTPCorsDomains = "http://localhost:7500,http://127.0.0.1:7500,http://0.0.0.0:7500,http://localhost:4242,http://127.0.0.1:4242,http://localhost:4200,http://0.0.0.0:4242,http://127.0.0.1:4200,http://0.0.0.0:4200,http://localhost:3000,http://0.0.0.0:3000,http://127.0.0.1:3000"`

File: cmd/validator/flags/flags.go (hunk 26 above)
Comment (james-prysm) [reply cmd/validator/flags/flags.go:118]: actually lets not do that, i looked at the references and it seems like it's just for a flag check. we don't need the default for anything else right now. we need a better strategy for these config values in general though ...

**Review by rkapka — APPROVED** [review]

File: api/server/httprest/options.go [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+	}
+}
+
+// WithTimeout allows changing the timeout value for API calls.
+func WithTimeout(seconds uint64) Option {
+	return func(g *Server) error {
+		g.cfg.timeout = time.Second * time.Duration(seconds)
+		return nil
+	}
+}
```
Comment (prestonvanloon) [inline api/server/httprest/options.go:34]: Why is this in seconds? Why not have the argument be time.Duration? Then it can be any unit of time, not just seconds.

File: api/server/httprest/server.go [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+	var handler http.Handler
+	if g.cfg.timeout > 10 {
+		handler = http.TimeoutHandler(g.cfg.router, g.cfg.timeout, "request timed out")
+	} else {
+		handler = g.cfg.router
+	}
+	g.server = &http.Server{
+		Addr:              g.cfg.httpAddr,
+		Handler:           handler,
+		ReadHeaderTimeout: time.Second,
```
Comment (prestonvanloon) [inline api/server/httprest/server.go:54]: Shouldn't this use the cfg.timeout?

**Review by saolyn — APPROVED** [review]
lgtm!

**Review by rkapka — APPROVED** [review]
